src/Raven.Server/Documents/Sharding/ShardExecutor.cs

```diff
@@ -0,0 +1,136 @@
+        {
+            var position = 0;
+            foreach (int shard in shards)
+            {
+                var cmd = _operation.CreateCommandForShard(shard);
+                _commands[position].Shard = shard;
+                _commands[position].Command = cmd;
+
+                var executor = _shardedContext.RequestExecutors[shard];
+                var release = executor.ContextPool.AllocateOperationContext(out JsonOperationContext ctx);
+                _commands[position].ContextReleaser = release;
+
+                var t = executor.ExecuteAsync(cmd, ctx);
+                _commands[position].Task = t;
+
+                if (executionMode == ExecutionMode.OneByOne)
+                {
+                    try
+                    {
+                        await t;
+                    }
+                    catch
+                    {
+                        if (failureMode == FailureMode.Throw)
+                            throw;
+                    }
+                }
+
+                position++;
+            }
+
+            for (var i = 0; i < position; i++)
+            {
+                var holder = _commands[i];
+                try
+                {
+                    await holder.Task;
+                }
+                catch (Exception e)
+                {
+                    if (failureMode == FailureMode.Throw)
+                        throw;
+
+                    _exceptions ??= new Dictionary<int, Exception>();
+                    _exceptions[holder.Shard] = e;
+                }
+            }
+
+            var resultsArray = ArrayPool<T>.Shared.Rent(position);
+            try
+            {
+                for (int i = 0; i < position; i++)
+                {
+                    resultsArray[i] = _commands[i].Command.Result;
+                }
+
+                return _operation.Combine(new Memory<T>(resultsArray, 0, position));
+            }
+            finally
+            {
+                ArrayPool<T>.Shared.Return(resultsArray);
+            }
+        }
+
+        public void Dispose()
+        {
+            if (_commands != null)
+            {
+                for (var index = 0; index < _commands.Length; index++)
+                {
+                    var command = _commands[index];
+                    try
+                    {
+                        command.ContextReleaser?.Dispose();
+                        command.ContextReleaser = null; // we set it to null, since we pool it and might get old values if not cleared
+                    }
+                    catch
+                    {
+                        // ignore
+                    }
+                }
+
+                ArrayPool<CommandHolder>.Shared.Return(_commands);
+            }
+        }
+
+        private struct CommandHolder
+        {
+            public int Shard;
+            public RavenCommand<T> Command;
+            public Task Task;
+            public IDisposable ContextReleaser;
```

Comment:
Why isn't this a `Scope`? We are boxing here, no?

Reply:
`IDisposable` is the return value of `AllocateOperationContext`